source/server/drain_manager_impl.cc

```diff
@@ -63,27 +71,40 @@ bool DrainManagerImpl::drainClose() const {
          (server_.api().randomGenerator().random() % drain_time_count);
 }
 
-Network::DrainDirection DrainManagerImpl::drainDirection() const { return draining_.load().second; }
-
 void DrainManagerImpl::startDrainSequence(Network::DrainDirection direction,
                                           std::function<void()> drain_complete_cb) {
-  ASSERT(!drain_tick_timer_);
-  drain_tick_timer_ = server_.dispatcher().createTimer(drain_complete_cb);
-
+  ASSERT(direction != Network::DrainDirection::None, "a valid direction must be specified.");
+  ASSERT(drain_tick_timers_.count(direction) == 0,
+         "cannot run two drain sequences for the same direction.");
+  auto new_timer = server_.dispatcher().createTimer(drain_complete_cb);
   const std::chrono::seconds drain_delay(server_.options().drainTime());
-  drain_tick_timer_->enableTimer(drain_delay);
+  new_timer->enableTimer(drain_delay);
+  drain_tick_timers_[direction] = std::move(new_timer);
   // Note https://github.com/envoyproxy/envoy/issues/31457, previous to which,
   // drain_deadline_ was set *after* draining_ resulting in a read/write race between
   // the main thread running this function from admin, and the worker thread calling
   // drainClose. Note that drain_deadline_ is default-constructed which guarantees
   // to set the time-since epoch to a count of 0
   // (https://en.cppreference.com/w/cpp/chrono/time_point/time_point).
-  ASSERT(drain_deadline_.time_since_epoch().count() == 0, "drain_deadline_ cannot be set twice.");
+  std::string direction_str;
+  switch (direction) {
+  case Network::DrainDirection::InboundOnly:
+    direction_str = "inbound";
+    break;
+  case Network::DrainDirection::All:
+    direction_str = "all";
+    break;
+  case Network::DrainDirection::None:
+    direction_str = "none";
+    break;
+  }
```

Comment:
seems like direction_str is not used?

Reply:
Whoops, this is some old debugging code; good catch